File: extension/src/main/java/io/holunda/camunda/bpm/data/factory/SetVariableFactory.java
package io.holunda.camunda.bpm.data.factory;

import io.holunda.camunda.bpm.data.adapter.ReadAdapter;
import io.holunda.camunda.bpm.data.adapter.WriteAdapter;
import io.holunda.camunda.bpm.data.adapter.set.SetReadWriteAdapterRuntimeService;
import io.holunda.camunda.bpm.data.adapter.set.SetReadWriteAdapterTaskService;
import io.holunda.camunda.bpm.data.adapter.set.SetReadWriteAdapterVariableMap;
import io.holunda.camunda.bpm.data.adapter.set.SetReadWriteAdapterVariableScope;
import org.camunda.bpm.engine.RuntimeService;
import org.camunda.bpm.engine.TaskService;
import org.camunda.bpm.engine.delegate.VariableScope;
import org.camunda.bpm.engine.variable.VariableMap;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Set;

/**
 * Variable factory of a base parametrized set type.
 *
 * @param <T> member type of the factory.
 */
public class SetVariableFactory<T> implements VariableFactory<Set<T>> {

  @NotNull
  private final String name;

  @NotNull
  private final Class<T> memberClazz;

  public SetVariableFactory(@NotNull String name, @NotNull Class<T> memberClazz) {
    this.name = name;
    this.memberClazz = memberClazz;
  }

  @Override
  public WriteAdapter<Set<T>> on(VariableScope variableScope) {
    return new SetReadWriteAdapterVariableScope<>(variableScope, name, memberClazz);
  }

  @Override
  public ReadAdapter<Set<T>> from(VariableScope variableScope) {
    return new SetReadWriteAdapterVariableScope<>(variableScope, name, memberClazz);
  }

  @Override
  public WriteAdapter<Set<T>> on(VariableMap variableMap) {
    return new SetReadWriteAdapterVariableMap<>(variableMap, name, memberClazz);
  }

  @Override
  public ReadAdapter<Set<T>> from(VariableMap variableMap) {
    return new SetReadWriteAdapterVariableMap<>(variableMap, name, memberClazz);
  }

  @Override
  public WriteAdapter<Set<T>> on(RuntimeService runtimeService, String executionId) {
    return new SetReadWriteAdapterRuntimeService<>(runtimeService, executionId, name, memberClazz);
  }

  @Override
  public ReadAdapter<Set<T>> from(RuntimeService runtimeService, String executionId) {
    return new SetReadWriteAdapterRuntimeService<>(runtimeService, executionId, name, memberClazz);
  }

  @Override
  public WriteAdapter<Set<T>> on(TaskService taskService, String taskId) {
    return new SetReadWriteAdapterTaskService<>(taskService, taskId, name, memberClazz);
  }

  @Override
  public ReadAdapter<Set<T>> from(TaskService taskService, String taskId) {
    return new SetReadWriteAdapterTaskService<>(taskService, taskId, name, memberClazz);
  }

  @Override
  @NotNull
  public String getName() {
    return name;
  }

  @NotNull
  public Class<T> getMemberClass() {
    return memberClazz;
  }

  @Override public boolean equals(Object o) {
    if (this == o)
      return true;
    if (o == null || getClass() != o.getClass())
      return false;
    SetVariableFactory<?> that = (SetVariableFactory<?>) o;
    return name.equals(that.name) &&
      memberClazz.equals(that.memberClazz);
  }

  @Override public int hashCode() {
    return Objects.hash(name, memberClazz);
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/factory/SetVariableFactory.kt
package io.holunda.camunda.bpm.data.factory

import io.holunda.camunda.bpm.data.adapter.ReadAdapter
import io.holunda.camunda.bpm.data.adapter.WriteAdapter
import io.holunda.camunda.bpm.data.adapter.set.*
import org.camunda.bpm.engine.CaseService
import org.camunda.bpm.engine.RuntimeService
import org.camunda.bpm.engine.TaskService
import org.camunda.bpm.engine.delegate.VariableScope
import org.camunda.bpm.engine.externaltask.LockedExternalTask
import org.camunda.bpm.engine.variable.VariableMap
import java.util.*

/**
 * Variable factory of a base parametrized set type.
 *
 * @param [T] member type of the factory.
 */
class SetVariableFactory<T>(override val name: String, val memberClass: Class<T>) : VariableFactory<Set<T>> {

  override fun on(variableScope: VariableScope): WriteAdapter<Set<T>> {
    return SetReadWriteAdapterVariableScope(variableScope, name, memberClass)
  }

  override fun from(variableScope: VariableScope): ReadAdapter<Set<T>> {
    return SetReadWriteAdapterVariableScope(variableScope, name, memberClass)
  }

  override fun on(variableMap: VariableMap): WriteAdapter<Set<T>> {
    return SetReadWriteAdapterVariableMap(variableMap, name, memberClass)
  }

  override fun from(variableMap: VariableMap): ReadAdapter<Set<T>> {
    return SetReadWriteAdapterVariableMap(variableMap, name, memberClass)
  }

  override fun on(runtimeService: RuntimeService, executionId: String): WriteAdapter<Set<T>> {
    return SetReadWriteAdapterRuntimeService(
      runtimeService,
      executionId,
      name,
      memberClass
    )
  }

  override fun from(runtimeService: RuntimeService, executionId: String): ReadAdapter<Set<T>> {
    return SetReadWriteAdapterRuntimeService(
      runtimeService,
      executionId,
      name,
      memberClass
    )
  }

  override fun on(taskService: TaskService, taskId: String): WriteAdapter<Set<T>> {
    return SetReadWriteAdapterTaskService(taskService, taskId, name, memberClass)
  }

  override fun from(taskService: TaskService, taskId: String): ReadAdapter<Set<T>> {
    return SetReadWriteAdapterTaskService(taskService, taskId, name, memberClass)
  }

  override fun on(caseService: CaseService, caseExecutionId: String): WriteAdapter<Set<T>> {
    return SetReadWriteAdapterCaseService(
      caseService,
      caseExecutionId,
      name,
      memberClass
    )
  }

  override fun from(caseService: CaseService, caseExecutionId: String): ReadAdapter<Set<T>> {
    return SetReadWriteAdapterCaseService(
      caseService,
      caseExecutionId,
      name,
      memberClass
    )
  }

  override fun from(lockedExternalTask: LockedExternalTask): ReadAdapter<Set<T>> {
    return SetReadAdapterLockedExternalTask(
      lockedExternalTask,
      name,
      memberClass
    )
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as SetVariableFactory<*>
    return name == that.name && memberClass == that.memberClass
  }

  override fun hashCode(): Int {
    return Objects.hash(name, memberClass)
  }

  override fun toString(): String {
    return "SetVariableFactory{" +
      "name='" + name + '\'' +
      ", memberClazz=" + memberClass +
      '}'
  }
}
